### PR TITLE
Rearrange docs TOC

### PR DESCRIPTION
**Pull request by mrocklin**
This makes a few changes to our top-level TOC.
Mostly this is motivated because I can never find some of our more
useful (by my estimation anyway) docs pages.

1.  Move deploying up towards the top in getting started
    (this used to be a frequently referenced page)
2.  Demote "How To" (which seems to be an odd grab-bag of topics without
    much organization currently) down to reference at the bottom
3.  Separate out the User APIs from the internal docs
    (I think that different folks are in different frames of mind for
    this)
4.  Promote Understanding Performance to top level (it used to be buried
    in the internals page, which I don't think was very visible)
5.  Promote Connect to remote data to top level (same)

@jacobtomlinson @jsignell

### CODE REVIEW

**Comment by jacobtomlinson** [comment]
I totally agree that things aren't in their perfect places yet, but this change feels a little like a revert on the overhaul we are working on. Perhaps it would be good to sync so we can chat about where we are trying to get to with the docs and see if there is a way we can surface the things you're struggling to find but stick to the ethos we are working to?

**Comment by mrocklin** [comment]
I'm happy to sync any time.  When works?

I don't think that we should pursue an ethos if we're not tracking if that ethos is actually having an impact.

These pages are, I think, very important for users to see.  I would challenge any ethos that would hide these pages.

**Comment by mrocklin** [comment]
Or rather I should probably turn my second line there into a question.  Do we have any information that these changes have improved reader experience?

If not, I'll provide my own experience, which is that I find it very hard to find pages that I find myself referring users to frequently.  Pages like setup and understanding performance answer a significant fraction (10s of percents) of questions that I see in the wild.  In converse, pages like the how-to index are, I think, full of not very useful pages in an order that I don't think makes sense.  If you try looking for setup or deploy the first page you often run into is how to deploy adaptive clusters.  This doesn't seem like the right flow to me.

File: docs/source/index.rst [hunk 1]
```diff
@@ -166,6 +173,7 @@ messy situations in everyday problems.
    develop.rst
    changelog.rst
    configuration.rst
+   how-to/index.rst
```
Comment (jacobtomlinson) [inline docs/source/index.rst:176]: A "how to" cookbook type section is one of the fundamental sections in Diataxis. Right now we don't have much of this content and it's become a bit of a catchall for stuff we haven't put in the right place yet. But as we go forwards I want to make more of this section, not less of it.

I don't think demoting this is the right move.

File: docs/source/index.rst [hunk 2]
```diff
@@ -151,10 +151,17 @@ messy situations in everyday problems.
    dataframe.rst
    delayed.rst
    futures.rst
+   how-to/connect-to-remote-data.rst
+
+.. toctree::
+   :maxdepth: 1
+   :hidden:
+   :caption: Internals
```
Comment (jacobtomlinson) [inline docs/source/index.rst:159]: The "Fundamentals" section is intended to be the "Explaination" section from Diataxis, and we have been trying to keen the number of sections limited to ease of exploring. However this one has grown a bit too big.

I think I'm in favour of this change to break it into two sections. What do you think @jsignell?

File: docs/source/index.rst [hunk 3]
```diff
@@ -136,9 +136,9 @@ messy situations in everyday problems.
 
    install.rst
    10-minutes-to-dask.rst
+   deploying.rst
```
Comment (jacobtomlinson) [inline docs/source/index.rst:139]: I don't think the current deployment docs should live in "Getting Started". For new users cluster deployment is going to be a second step and more advanced topic. I think I would prefer for this to stay under the "Fundamentals" heading.

File: docs/source/index.rst [hunk 4]
```diff
@@ -151,10 +151,17 @@ messy situations in everyday problems.
    dataframe.rst
    delayed.rst
    futures.rst
+   how-to/connect-to-remote-data.rst
```
Comment (jacobtomlinson) [inline docs/source/index.rst:154]: This should stay in the "How to" section.

File: docs/source/index.rst [hunk 5]
```diff
@@ -151,10 +151,17 @@ messy situations in everyday problems.
...
+
+.. toctree::
+   :maxdepth: 1
+   :hidden:
+   :caption: Internals
+
+   understanding-performance.rst
```
Comment (jacobtomlinson) [inline docs/source/index.rst:161]: I think this makes sense.

**Review by jacobtomlinson — COMMENTED** [review]
I think it's important to note that the way you use the docs and direct folks around it is going to be different to the way new users will use it. We are trying to structure things in line with the [Diataxis framework](https://diataxis.fr/) as we are pretty sold on that being a useful organisational structure for the majority of users.

I've commented some direct feedback on the proposed changes, I'm about 50/50 with them.

I would be keen for @jsignell to weigh in too.

**Comment by mrocklin** [comment]
Merging at end of week if there are no further comments

**Comment by jacobtomlinson** [comment]
@mrocklin did you see my review?

File: docs/source/index.rst (hunk 3 above)
Comment (jsignell) [reply docs/source/index.rst:139]: I tend to agree but I can imagine the 10 minutes to dask doing a much better job of introducing deployment options

File: docs/source/index.rst (hunk 2 above)
Comment (jsignell) [reply docs/source/index.rst:159]: Yeah Internals does not seems like an intuitive name to me though. Something like "Deepening Understanding" not that, but something like that :smile:

File: docs/source/index.rst (hunk 1 above)
Comment (jsignell) [reply docs/source/index.rst:176]: I think the main value of the how-to is to be able to point directly to snippets. The list of all the pages is not as interesting to me. So I'd actually be fine just removing it from the toc for the time being and trying to plump it up a little before we promote it.

**Review by jsignell — COMMENTED** [review]
Sorry for being delinquent here. I agree with Jacob that none of us are the typical user of the docs. 

Also I am not surprised Matt, that you don't feel like you are experiencing the benefits of the new framework. I think that's because we haven't gone far enough yet. We did an initial push and have now left the docs in this kind of partial state for longer than intended. 

Maybe Jacob and I can put some time on the calendar to do some more work here and then we can do a few user tests or something to see how the docs feel to people who haven't stared at them for so long :)

File: docs/source/index.rst (hunk 1 above)
Comment (mrocklin) [reply docs/source/index.rst:176]: I think that if we have anything at this high level of visibility then it needs to be high quality.   This page is not currently high quality.  I think that we should remove this and then invest in it and then bring it back once that investment is done.

File: docs/source/index.rst (hunk 3 above)
Comment (mrocklin) [reply docs/source/index.rst:139]: This page is, I think, of very high value, especially for folks who are coming into Dask from a laptop perspective, but from a "I have a big dataset to deal with" perspective.  Right now if I go to the docs with that question in mind I don't see anything that I can click on that will take me closer to that information.

Comment (mrocklin) [reply docs/source/index.rst:139]: > I tend to agree but I can imagine the 10 minutes to dask doing a much better job of introducing deployment options

I'm not seeing deployment options in that page?

I think that people arrive at docs asking themselves questions like "How do I run this thing on Kubernetes?".  I don't think that we currently serve these readers.

**Comment by mrocklin** [comment]
> We did an initial push and have now left the docs in this kind of partial state for longer than intended.

Yeah, I think that our docs are important enough that we probably can't leave things in a half-done state.  If there is some grand architecture that we're doing then awesome.  Let's do it.  I'm going to push against merging that into main until it's ready though.  The How-To section seems very half-done.  I think that we should remove it until it is done and a lot of thought has been put into it.

> and then we can do a few user tests or something to see how the docs feel to people who haven't stared at them for so long :)

Huge :+1: to this if you can make the time for this.  I think that this would be very valuable.

**Comment by jacobtomlinson** [comment]
> Im going to push against merging that into main until it's ready though.

...

>  I think that we should remove it until it is done and a lot of thought has been put into it.

I'm a little prickly to these comments. A lot of time and effort has gone into the work already. I agree we need to put more time in but I generally think things are better than they were before. If other folks also want to help then that's great, but I'm keen to steer changes like this towards out goal, not reverting away from it.

File: docs/source/index.rst (hunk 3 above)
Comment (jsignell) [reply docs/source/index.rst:139]: Yeah there is a mention under "Scheduling" but to your point maybe that isn't framed in the way that people are asking it.

Comment (mrocklin) [reply docs/source/index.rst:139]: Yeah, as someone who knows Dask really well and who goes to the docs regularly I haven't yet successfully navigated to the deploy page.  It isn't natural to me to look at scheduling (which I view as a more technical topic) in order to learn how to set it up.  I would expect this more under something like "install".

**Comment by mrocklin** [comment]
> I'm a little prickly to these comments. A lot of time and effort has gone into the work already. I agree we need to put more time in but I generally think things are better than they were before. If other folks also want to help then that's great, but I'm keen to steer changes like this towards out goal, not reverting away from it

Sorry to be perhaps overly brusque.  I should get better at that.

Let me focus on the "How to" page.  It seems like a grab-bag of topics that isn't in any particular order.  I'm proud of lots of our doc pages, but I'm not currently proud of that one.  I don't think that it rises to the quality that is typical for Dask documentation.  I'd love to see it get there, but until it does I don't think tha we should promote it as prominently as we do today.

> A lot of time and effort has gone into the work already

Yeah, clearly it's great.  However I think that there are a few reasonable constraints that we're not hitting currently.  I'm fine moving in a direction promoted by the ethos, but I do think that a few questions need to be answerable at top level:

1. How do I run this thing on my cluster?
2. How do I access data?

I challenge any ethos that wouldn't put these things front and center.  I strongly disagree with any ethos that promotes educating people about adaptive deployments before educating them about just generally deploying the thing.


Maybe the question is "if we like this direction great, how do we serve common user questions while we build towards this direction?"


Separately, I would also like for us to question if this change in direction is good.  I think that we should be testing ourselves.  For example, if we saw time-on-site increase did we see time-on-page

**Comment by mrocklin** [comment]
Ping

**Comment by jacobtomlinson** [comment]
I feel like @jsignell and I have both reviewed this PR with some gentle pushback along the lines of "we are trying something here, and we're not done, could you make some small adjustments to accommodate our experiment". 

I completely agree that the docs today are not as optimal as we would like, and this change shuffles some things in a way that is great. But as I said already I want to try and iterate towards our goal rather than revert away from it. I also understand that both of our attentions are elsewhere right now, but my intention is to return to this activity soon.

I see us having two options here:
- You can choose to ignore our pushback and go ahead and merge.
- You can make some changes as per the reviews and we can merge.

File: docs/source/index.rst (hunk 2 above)
Comment (mrocklin) [reply docs/source/index.rst:159]: Any specific suggestions on the title here?

File: docs/source/index.rst (hunk 1 above)
Comment (mrocklin) [reply docs/source/index.rst:176]: It seems like @jsignell is ok with removing this until it becomes higher quality.  Leaving this in for now.

File: docs/source/index.rst (hunk 4 above)
Comment (mrocklin) [reply docs/source/index.rst:154]: I'll remove this for now, although I think that it's a mistake to not keep this top-level.  I think that "How do I access data?" should be a question that someone should be able to easily answer when looking at the docs.

**Comment by mrocklin** [comment]
Thanks for the response @jacobtomlinson !

> You can make some changes as per the reviews and we can merge

I'm not seeing a ton of actionable feedback in the reviews.  I've made one change and pushed it.  I've resolved conversations that I think are resolvable, and asked for insight on another.  Is there anything else I can do here to move this forward?

>  "we are trying something here, and we're not done"

My perspective here is that we've started something, are not done, but don't seem to have active plans to continue the effort.  If that's the case then I think that we should release constraints here and let others engage with other ideas.  However if I'm wrong here and you all are actively working on docs then I'll change my approach entirely.  If you all haven't moved on then I'll happily raise an issue with ideas/concerns/suggestions and engage you all there.  Thoughts?

**Comment by jacobtomlinson** [comment]
I think the main actions were I would like the "how to" location to be restored to "getting started", and I think "deploying" should live under "fundamentals".

> My perspective here is that we've started something, are not done, but don't seem to have active plans to continue the effort. If that's the case then I think that we should release constraints here and let others engage with other ideas. However if I'm wrong here and you all are actively working on docs then I'll change my approach entirely.

I consider @jsignell and I to be actively developing this. Things have just been a little slow as this is a back burner activity for us.

Given our resource constraints and your motivation to see more progress here perhaps we could leverage some of Coiled's advocacy/copywriting resources to boost things?

**Comment by mrocklin** [comment]
>  "how to" location to be restored

This page doesn't meet my standards for being included in the documentation, much less the top-level TOC.  I'm also going to lean on this suggestion from Julia:

>> I'd actually be fine just removing it from the toc for the time being and trying to plump it up a little before we promote it.

> and I think "deploying" should live under "fundamentals".

Sure.  Done.   Mostly I want this to be top-level rather than buried.

> I consider @jsignell and I to be actively developing this. Things have just been a little slow as this is a back burner activity for us.

I'm not seeing any major work on this this year (I stopped looking at that point).  Maybe I'm missing something though?  This feels abandoned to me.  Totally ok.  Everyone has a lot to do.  I do think that this has gone on long enough that it makes sense to release the lock on this effort though.

> Given our resource constraints and your motivation to see more progress here perhaps we could leverage some of Coiled's advocacy/copywriting resources to boost things

Certainly motivated to invest resources into documentation.  It's not immediately clear to me that this framework is better than other frameworks though.

**Comment by jacobtomlinson** [comment]
> I do think that this has gone on long enough that it makes sense to release the lock on this effort though.

Consider the lock released.

> Certainly motivated to invest resources into documentation. It's not immediately clear to me that this framework is better than other frameworks though.

I would be very happy for others to step in and think about this. The more the merrier.

We had a bunch of conversations about docs frameworks towards the end of last year. Diataxis seemed like a good fit and we put in some effort to try it out but got sidetracked before proving it was a good move. I still think it is a valid option and I would love to continue down this path to find out.

Equally, if others want to step in here but would prefer to take a different path then great, I'm happy to help out where I can. However, I do feel it's a little premature to abandon this approach.

**Comment by mrocklin** [comment]
I'm happy to continue to have our strategy be informed by a framework, but
I don't think that we should restrict ourselves to a framework if no one is
actively pushing it.

If someone proposes a change that doesn't fit into a framework but is
otherwise valuable (in terms of increasing engagement, answering very
common questions, etc.) then I wouldn't object to making that change at
this point.  This seems to be where you and I might disagree.

**Comment by mrocklin** [comment]
Woot.  Thanks Jacob!

**Pull request merged**